File: Java/opentraceability/src/main/java/opentraceability/utility/Countries.java
package opentraceability.utility;

import java.io.*;
import java.util.*;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import org.w3c.dom.*;
import org.xml.sax.InputSource;

public class Countries {

    public static Map<String, Country> _dirCountries = new HashMap<>();
    public static Map<String, Country> _dirAlpha3Countries = new HashMap<>();
    public static Map<String, Country> _dirNameCountries = new HashMap<>();

    public Countries() throws Exception {
        load();
    }

    public static void load() throws Exception {
        String data = StaticData.readData("/Countries.xml");
        _dirCountries = new HashMap<>();
        _dirNameCountries = new HashMap<>();
        _dirAlpha3Countries = new HashMap<>();
        try {
            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            InputSource xmlInput = new InputSource(new StringReader(data));
            Document xmlCountries = dBuilder.parse(xmlInput);

            Node root = xmlCountries.getDocumentElement();

            for (int i = 0; i < root.getChildNodes().getLength(); i++) {
                Node node = root.getChildNodes().item(i);

                if (node.getNodeType() == Node.ELEMENT_NODE) {
                    Element x = (Element) node;
                    Country country = new Country(x);
                    _dirCountries.put(country.abbreviation.toUpperCase(), country);
                    _dirNameCountries.put(country.name.toUpperCase(), country);

                    if (!country.alpha3.isEmpty()) {
                        _dirAlpha3Countries.put(country.alpha3.toUpperCase(), country);
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static List<Country> getCountryList() {
        List<Country> list = new ArrayList<>();

        _dirNameCountries.forEach((k, v) -> list.add(v));
        return list;
    }

    public static Country fromAbbreviation(String code) {
        Country country = null;

        if (!code.isEmpty()) {
            if (_dirCountries != null) {
                if (_dirCountries.containsKey((code.toUpperCase()))) {
                    country = _dirCountries.get(code.toUpperCase());
                }
            }
        }

        return country;
    }

    public static Country fromAlpha3(String code) {
        Country country = null;

        if (!code.isEmpty()) {
            if (_dirAlpha3Countries != null) {
                if (_dirAlpha3Countries.containsKey((code.toUpperCase()))) {
                    country = _dirAlpha3Countries.get(code.toUpperCase());
                }
            }
        }

        return country;
    }

    public static Country fromCountryName(String name) {
        Country country = null;

        if (_dirCountries != null && !name.isEmpty()) {

            if (_dirNameCountries.containsKey((name.toUpperCase()))) {
                country = _dirNameCountries.get(name.toUpperCase());
            } else {
                loop:
                for (Map.Entry<String, Country> element : _dirNameCountries.entrySet()) {
                    if (element.getValue().alternativeName != null) {
                        if (element.getValue().alternativeName.equalsIgnoreCase(name)) {
                            country = element.getValue();
                            break;
                        }
                    }
                }
            }
        }

        return country;
    }

    public static Country fromCountryIso(int iso) {
        Country country = null;

        if (_dirCountries != null) {

            loop:
            for (Map.Entry<String, Country> element : _dirCountries.entrySet()) {
                if (element.getValue().iso == iso) {
                    country = element.getValue();
                    break;
                }
            }

            return country;
        }

        return country;
    }

    public static Country parse(String strValue) {
        try {
            Integer parsedInt = Integer.parseInt(strValue);
            return fromCountryIso(parsedInt);
        }
        catch (Exception ex)
        {
            return fromAbbreviation(strValue) != null ? fromAbbreviation(strValue) : fromAlpha3(strValue) != null ? fromAlpha3(strValue) : fromCountryName(strValue);
        }
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/utility/countries.kt
package utility

import java.util.*
import javax.xml.parsers.DocumentBuilderFactory
import org.xml.sax.InputSource
import java.io.StringReader
import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node

class Countries {

    constructor() {
        _dirCountries = mutableMapOf()
        _dirAlpha3Countries = mutableMapOf()
        _dirNameCountries = mutableMapOf()
        load()
    }

    companion object {
        internal var _dirCountries: MutableMap<String, Country> = mutableMapOf()
        internal var _dirAlpha3Countries: MutableMap<String, Country> = mutableMapOf()
        internal var _dirNameCountries: MutableMap<String, Country> = mutableMapOf()

        fun load() {
            var data: String? = StaticData.readData("Countries.xml")

            val dbFactory = DocumentBuilderFactory.newInstance()
            val dBuilder = dbFactory.newDocumentBuilder()
            val xmlInput = InputSource(StringReader(data))
            val xmlCountries: Document = dBuilder.parse(xmlInput)

            val root = xmlCountries.documentElement // get root element

            for (i in 0 until root.childNodes.length) {
                val node = root.childNodes.item(i)
                if (node.nodeType == Node.ELEMENT_NODE) {
                    val x = node as Element
                    val country = Country(x)
                    _dirCountries[country.Abbreviation.toUpperCase()] = country
                    _dirNameCountries[country.Name.toUpperCase()] = country
                    if (!country.Alpha3.isNullOrEmpty()) {
                        _dirAlpha3Countries[country.Alpha3.toUpperCase()] = country
                    }
                }
            }
        }


        var CountryList: ArrayList<Country> = ArrayList<Country>()
            get() {
                var list: ArrayList<Country> = ArrayList()

                _dirNameCountries.forEach { element ->
                    list.add(element.value)
                }
                return list
            }


        fun fromAbbreviation(code: String): Country? {
            var country: Country? = null

            if (!code.isNullOrEmpty()){
                if (_dirCountries != null){
                    if (_dirCountries.containsKey((code.toUpperCase()))){
                        country = _dirCountries.get(code.toUpperCase())!!;
                    }
                }
            }

            return country
        }

        fun fromAlpha3(code: String): Country? {
            var country: Country? = null

            if (!code.isNullOrEmpty()){
                if (_dirCountries != null){
                    if (_dirAlpha3Countries.containsKey((code.toUpperCase()))){
                        country = _dirCountries.get(code.toUpperCase())!!;
                    }
                }
            }

            return country
        }

        fun fromCountryName(name: String): Country? {
            var country: Country? = null

            if (_dirCountries != null && !name.isNullOrEmpty()){

                if (_dirNameCountries.containsKey((name.toUpperCase()))){
                    country = _dirCountries.get(name.toUpperCase())!!;
                }
                else{
                    run loop@{
                        _dirNameCountries.forEach { element ->
                            if (element.value.AlternativeName != null) {
                                if (element.value.AlternativeName.toUpperCase() == name.toUpperCase()) {
                                    country = element.value;
                                    return@loop
                                }
                            }
                        }
                    }
                }
            }

            return country
        }

        fun fromCountryIso(iso: Int): Country? {
            var country: Country? = null

            if (_dirCountries != null){

                run loop@{
                    _dirCountries.forEach { element ->
                        if (element.value.ISO == iso) {
                            country = element.value;
                            return@loop
                        }
                    }
                }

                return country
            }

            return country
        }

        fun parse(strValue: String): Country? {

            val parsedInt = strValue.toIntOrNull()
            if (parsedInt != null) {
                return fromCountryIso(parsedInt);
            } else {
                return fromAbbreviation(strValue) ?: fromAlpha3(strValue) ?: fromCountryName(strValue)
            }
        }
    }
}
